### DISCUSSION TITLE
Transaction aware storage provider

### DISCUSSION THREAD

**Discussion by zhemaituk** [jobrunr/jobrunr#279] [Ideas]
For me it's a common use case to schedule a job in a @Transactional context (spring-boot) and do some other persistent operations at the same time.
Something like this:

```
    @Transactional
    public Entity saveAndLaunch(Entity details) {
        final Entity result = repo.save(details);
        jobScheduler.enqueue(result.getId(), () -> jobProcessor.execute(result.getId()));
        return result;
    }
```

It was somewhat a surprise for me that jobScheduler.enqueue under the hood created a separate transaction and committed it.
So in rare cases when the job starts processing before the "parent" @Transactional method completes and commits own transaction -> the job fails with EntityNotFoundException.

@rdehuyss 
Does it make sense for JobRunr to detect that it's running in a transactional environment already, participate in Spring Managed Transaction, and let the caller/spring to commit the transaction in such case?

**Comment by rdehuyss** [comment]
You can use `TransactionAwareDataSourceProxy` from Spring.

**Comment by zhemaituk** [comment]
Thanks, it helped to overcome EntityNotFoundException, though brought me some exciting troubleshooting times :)

What's happening:
DefaultSqlStorageProvider does commit transaction under the hood.
So still the job start executing before parent @Transactional method completes.
And if something happens after the enqueue - the changes are not rolled back. Including the partial changes of the parent transaction.

```
    @Transactional
    public Entity saveAndLaunch(Entity details) {
        final Entity result = repo.save(details);
        jobScheduler.enqueue(result.getId(), () -> jobProcessor.execute(result.getId()));
        
        // Transaction committed, so the executor picks up the job in a parallel thread. It does not fail with EntityNotFoundException anymore, as Entity was committed as well.
        Thread.sleep(10000);
        
        // Further logic goes here.

        // If something bad happens, only the persistent changes after enqueue are rolled back, potentially leaving database in logically inconsistent state
        throw new RuntimeException("Something bad happened");
    }
```

Would be ideal for the implementation not to commit if it executes in a transactional context, but leave transaction management to the caller.

Less important, but jobrunr-spring-boot-starter could auto-configure for transaction awareness, as it knows already it runs in spring boot environment.

**Comment by zhemaituk** [comment]
@rdehuyss what do you think about not force-committing transaction in DefaultSqlStorageProvider?

**Reply by rdehuyss** [reply]
Sorry, can't do - JHipster needs this.

**Comment by v13leo** [comment]
I see this is an old thread, but just want to bump and second this. I'm using postgres with the spring-boot-starter and I also have a scenario where a parent transaction can rollback without the enqueue process also rolling back.

It'd be great if the caller of the enqueue had some control over the committing process or if it maybe used 
https://docs.oracle.com/javaee/7/api/javax/transaction/TransactionManager.html to determine if it's currently operating within a transaction and can expect the commit to happen elsewhere

**Reply by rdehuyss** [reply]
I looked into https://docs.oracle.com/javaee/7/api/javax/transaction/TransactionManager.html  - easier said than done :-), especially as I want to stay backwards compatible with other frameworks that do need explicit committing.

I also don't like to obfuscate the JobRunr Api to allow control over the committing - that's against the moto of JobRunr (easy background job processing). I do understand the need though...

**Comment by rdehuyss** [comment]
Good news: this will be part of **JobRunr Pro 5.0** where any job that is created (`enqueued` or `scheduled`) will out-of-the-box take part in Spring managed transactions (so all methods annotated with Spring's `@Transactional` annotation).

The difficulty was to keep it compatible with all frameworks, including Spring, Micronaut and JHipster.

**Reply by zhemaituk** [reply]
Thank you Ronald,

Looking forward to version 5!

If anyone stumbles upon similar, temp solution for v4 I ended up with.
In `application.yaml`:
```
org:
  jobrunr:
    database:
      datasource: transactionAwareDataSource
```

`DataSourceConfig.java`:
```
@Configuration
public class DataSourceConfig {

    @Primary
    @Bean
    @ConfigurationProperties(prefix = "spring.datasource")
    public DataSource dataSource(DataSourceProperties properties) {
        return properties.initializeDataSourceBuilder().build();
    }

    @Bean(name = "transactionAwareDataSource")
    public DataSource transactionAwareDataSource(DataSource dataSource) {
        return dataSource instanceof TransactionAwareDataSourceProxy ? dataSource : new TransactionAwareDataSourceProxy(dataSource);
    }

}
```

In the code where enqueue happens:
```
final MyEntity result = repository.save(entity);
// Flushing changes to MyEntity first, as Spring may delay it till returning from @Transactional method
getEntityManager().flush();
// WARN: enqueue operation commits the transaction (https://github.com/jobrunr/jobrunr/discussions/279)
jobScheduler.enqueue(result.getId(), () -> backtestJobProcessor.execute(result.getId()));
return result;
```
